src/test/java/jenkins/branch/harness/MultiBranchImpl.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.branch.harness;
+
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.util.List;
+
+import hudson.Extension;
+import hudson.model.FreeStyleBuild;
+import hudson.model.FreeStyleProject;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.TopLevelItem;
+import hudson.scm.SCM;
+import hudson.scm.SCMDescriptor;
+import jenkins.branch.Branch;
+import jenkins.branch.BranchProjectFactory;
+import jenkins.branch.MultiBranchProject;
+import jenkins.branch.MultiBranchProjectDescriptor;
+
+public class MultiBranchImpl extends MultiBranchProject<FreeStyleProject, FreeStyleBuild> {
+
+
+    protected MultiBranchImpl(ItemGroup parent, String name) {
+        super(parent, name);
+    }
+
+    @Override
+    protected BranchProjectFactory<FreeStyleProject, FreeStyleBuild> newProjectFactory() {
+        return new BranchProjectFactoryImpl();
+    }
+
+    /*
+    @Override
+    public List<SCMSource> getSCMSources() {
+        if (getSourcesList() == null) {
+            // test code is generating a NullPointer when calling it from an ItemListener on the SCMSourceOwner
+            // It seems that the object is not fully initialized when the ItemListener uses it.
+            // Perhaps it needs to be reproduced and investigated in a branch-api test.
+            return new ArrayList<SCMSource>();
+        }
+        return super.getSCMSources();
+    }
+    */
+
+    public static class BranchProjectFactoryImpl extends BranchProjectFactory<FreeStyleProject, FreeStyleBuild> {
+
+        @Override
+        public FreeStyleProject newInstance(Branch branch) {
+            FreeStyleProject job = new FreeStyleProject(getOwner(), branch.getName());
+            job.onCreatedFromScratch();
```

Comment:
:bug: this is not to be called by `newInstance` implementations as per documentation.